### ISSUE TITLE
Training Large-Scale Tesseract OCR Model for Uchen Script

### ISSUE THREAD

**Issue by tenzinyonten** [OpenPecha/Tesseract-Uchen-Training#1]
**Description**

 Train Tesseract OCR model on multiple large-scale Uchen script datasets (~4M images) for improved Tibetan text recognition.

**Implementation:**

2.	Data preprocessing and format conversion
3.	Training file generation
4.	Model training and optimisation

**Subtasks:**

- [x] Consolidate all datasets and verify completeness 
- [x] Convert text files to ground truth format
- [x] Configure Makefile parameters
- [x] 	Generate box and lstmf files
- [ ] 	Execute and monitor training process

**Completion Criteria: **

Successfully train a Tesseract model for Uchen script recognition

**Card Reviewer: **

- [x] @ta4tsering

**Comment by tenzinyonten** [comment]
The Karmapa8 dataset currently doesn’t have a predefined data split. I reached out to @ta4tsering , and he confirmed he will prepare it and provide it by today.

**Comment by ta4tsering** [comment]
I have updated the blog as well as the hugging face with three new repos of Uchan data with data distribution or the data split. 

https://huggingface.co/datasets/openpecha/OCR-NorbuketakaNumbers
https://huggingface.co/datasets/openpecha/OCR-Karmapa8
https://huggingface.co/datasets/openpecha/OCR-KhyentseWangpo

**Comment by kaldan007** [comment]
@tenzinyonten can u elaborate the step number 4

**Comment by TenYonten** [comment]
@kaldan007 Expanding on the task card - here's the detailed LSTM workflow in Tesseract 5.0+:

For data preparation:

- Input: line-segmented Uchen script images
- Ground truth: Name files as .gt.txt (e.g., line_01.gt.txt), ensure exact text matching, and use UTF-8 encoding
- Tesseract 5 internally handles: image normalization, feature extraction, character encoding, and line geometry

Training files are automatically generated:

1. Box files (line-level, not character-level):
     Analyzes image regions
     Maps line coordinates and text content
     Handles alignment information
2. LSTMF (LSTM Feature) files - crucial for efficient training:
     **Optimization:** Convert raw images and alignments into a binary format to avoid redundant computations.
     **Integration:** Combine preprocessed image features, box data, and ground truth into a compact package.

LSTMF files serve as preprocessed and optimized representations of training data, integrating image features, text alignments, and ground truth into a compact binary format. This ensures faster training by reducing redundant computations.

Makefile config needed:

```
START_MODEL= (blank for fresh training). 
#If START_MODEL is blank, training starts from scratch. Otherwise, specify a pre-trained model (e.g., 'eng' or 'tib') to fine-tune on your dataset.
MODEL_NAME=tib
DATA_DIR=/path/to/training/data
TESSDATA=/path/to/tessdata
MAX_ITERATIONS=2000000
LEARNING_RATE=0.001
LANG_TYPE=Indic  
 #Set LANG_TYPE=Indic for scripts like Tibetan that use complex ligatures and require advanced shaping logic during recognition.

```

Training Flow:
Input: Line Images + .gt.txt.
Internal Processing: Tesseract handles image normalization, feature extraction, and alignment preparation.
LSTM Training: The preprocessed data is fed into the LSTM network for recognition training.


![Image](https://github.com/user-attachments/assets/c0d9a8a3-d8ae-4f8a-9d07-b4013d6c3366)

**Comment by kaldan007** [comment]
the suggested data sample for parallel training is:
30k,300k,3mill

**Comment by tenzinyonten** [comment]
**we have 6 datasets**
### Modern books
 
Norbu
 
![Image](https://github.com/user-attachments/assets/58ce0caa-2e88-4821-957f-bb38c59c532f)
 
Google books
 
![Image](https://github.com/user-attachments/assets/0c5869a6-fff9-4d7a-b6e4-775b7c20df84)
 
 
### Manuscript
Karmapa
 
 
![Image](https://github.com/user-attachments/assets/e459064b-0936-4d44-ac8e-fe283bf928c8)
 
Derge
 
 
![Image](https://github.com/user-attachments/assets/08f0ead3-4a2c-42d1-ae2a-733065b081c6)
 
Lhasa
 
 
![Image](https://github.com/user-attachments/assets/fd6e63ff-c05c-4a30-a33e-ba0908e18168)
 
 
Lithang
 
![Image](https://github.com/user-attachments/assets/df3cc552-2074-45e5-bd52-3952c018e552)
 
### **Training File Details**
 
**Total training files**: 3,688,354
 
**Dataset proportions**: 
   - **Norbuketaka**: 48.60% 
   - **Derge**: 18.95% 
   - **Google Books**: 16.42% 
   - **Lithang**: 11.90% 
   - **Lhasa**: 3.53% 
   - **Karmapa**: 0.60% 
 
**Subset Distribution Plan**: 
| Subset   | Total Images | Norbuketaka | Derge   | Google Books | Lithang | Lhasa  | Karmapa |
|----------|--------------|-------------|---------|--------------|---------|--------|---------|
| **Small**  | 37,000       | 17,982      | 7,001   | 6,312        | 4,393   | 1,307  | 222     |
| **Medium** | 370,000      | 179,220     | 69,965  | 60,748       | 43,970  | 13,071 | 2,220   |
| **Full**   | 3,688,354    | 1,792,364   | 699,019 | 605,602      | 438,999 | 130,109| 22,261  

Proportional dataset subsets for **37K**, **370K**, and **full dataset (3.7M)**.

**Comment by tenzinyonten** [comment]
## Dataset Configurations
1. Small Dataset (37K samples)
   - Max Iterations: 300,000
   - Learning Rate: 0.002
   - Language Type: Indic
   - Page Segmentation Mode (PSM): 13


2. Medium Dataset (370K samples)
   - Max Iterations: 1,000,000
   - Learning Rate: 0.0005
   - Language Type: Indic
   - Page Segmentation Mode (PSM): 13

3. Large Dataset (3.7M samples)
   - Max Iterations: 10,000,000
   - Learning Rate: 0.0001
   - Language Type: Indic
   - Page Segmentation Mode (PSM): 13


Training Results - Tesseract LSTM (tib_small):
Best BCER: 5.03%

**Comment by kaldan007** [comment]
@tenzinyonten what was the time estimation to train for 3.7 m samples.

**Comment by tenzinyonten** [comment]
@kaldan007  It took about 2.5 days to train on 370K samples, but our full dataset of 3.7M samples will likely take about 2.5-3 weeks.
